iota-client/src/node/mqtt.rs

```diff
@@ -0,0 +1,207 @@
+
+macro_rules! lazy_static {
+  ($init:expr => $type:ty) => {{
+    static mut VALUE: Option<$type> = None;
+    static INIT: std::sync::Once = std::sync::Once::new();
+
+    INIT.call_once(|| unsafe { VALUE = Some($init) });
+    unsafe { VALUE.as_ref() }.expect("failed to get lazy static value")
+  }};
+}
+
+/// A topic.
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub struct Topic(String);
+
+impl TryFrom<&str> for Topic {
+  type Error = crate::Error;
+
+  fn try_from(value: &str) -> std::result::Result<Self, Self::Error> {
+    Self::new(value)
+  }
+}
+
+impl Topic {
+  /// Creates a new topic and checks if it's valid.
+  pub fn new<S: Into<String>>(name: S) -> Result<Self> {
+    let valid_topics = lazy_static!(
+      ["milestones/latest", "milestones/solid", "messages", "messages/referenced"].to_vec() => Vec<&str>
+    );
+    let regexes = lazy_static!(
+      [
+        Regex::new(r"messages/([A-Fa-f0-9]{64})/metadata").unwrap(),
+        Regex::new(r"outputs/([A-Fa-f0-9]{64})(\d{4})").unwrap(),
+        Regex::new("addresses/([A-Fa-f0-9]{64})/outputs").unwrap(),
+        Regex::new(r"messages/indexation/(\.)").unwrap()
+      ].to_vec() => Vec<Regex>
+    );
+    let name = name.into();
+    if valid_topics.iter().any(|valid| valid == &name)
+      || regexes.iter().any(|re| re.is_match(&name))
+    {
+      let topic = Self(name);
+      Ok(topic)
+    } else {
+      Err(crate::Error::InvalidMqttTopic(name))
+    }
+  }
+}
+
+pub(crate) fn get_mqtt_client(client: &mut Client) -> Result<&MqttClient> {
+  match client.mqtt_client {
+    Some(ref c) => Ok(c),
+    None => {
+      for node in client.pool.read().unwrap().iter() {
+        // node.set_path("mqtt");
+        let uri = &format!(
+          "{}://{}:{}/mqtt",
+          if node.scheme() == "https" {
+            "wss"
+          } else {
+            "ws"
+          },
+          node.host_str().unwrap(),
+          node.port_or_known_default().unwrap()
+        );
+        let mqtt_options = CreateOptionsBuilder::new()
+          .server_uri(uri)
+          .client_id(format!("iota.rs-mqtt-client-{}", uri))
+          .finalize();
+        let mut mqtt_client = MqttClient::new(mqtt_options)?;
+
+        let conn_opts = ConnectOptionsBuilder::new()
+          .keep_alive_interval(Duration::from_secs(20))
+          .mqtt_version(MQTT_VERSION_3_1_1)
+          .clean_session(true)
+          .finalize();
+
+        if let Ok(_) = mqtt_client.connect(conn_opts) {
+          poll_mqtt(client.mqtt_topic_handlers.clone(), &mut mqtt_client);
+          client.mqtt_client = Some(mqtt_client);
+          break;
+        }
+      }
+      client
+        .mqtt_client
+        .as_ref()
+        .ok_or_else(|| crate::Error::MqttConnectionNotFound)
+    }
+  }
+}
+
+fn poll_mqtt(mqtt_topic_handlers: Arc<Mutex<TopicHandlerMap>>, client: &mut MqttClient) {
+  let receiver = client.start_consuming();
+  std::thread::spawn(move || {
+    while let Ok(message) = receiver.recv() {
```

Comment:
Can you help check the while let loop will break when the all the topics are unsubscribed? (Like the mpsc Receiver will be dropped.) And it might be good to handle the corresponding Sender is disconnected.

Reply:
Why I point this out is that I saw this [issue](https://github.com/eclipse/paho.mqtt.rust/issues/61).

Reply:
do you think we should close the connection when all the topics are unsubscribed? for performance reasons I think we could expose a method to do it, because for instance the wallet will unsubscribe to all topics and then subscribe again frequently (on every sync).

Reply:
If we will subscribe/unsubscribe frequently I think it's good to remain the connection. Just we might need to expose a method to let the user disconnect and let the Receiver channel to be dropped, so as to broke the `while let` loop and let the spawned thread be shut down gracefully. Or do we already have a way to let the spawned thread gracefully shut down if the user does not want to receive mqtt messages?

Reply:
just pushed the changes for it, yeah I had to drop the mqtt client instance in order to break from the while loop and shut down the thread. thanks for pointing that out, nice catch :P

Reply:
I also added a `broker_options` field to the client builder with a flag for the automatic disconnection when the subscriptions are empty (defaults to true), is that ok @fijter ?